File: solutions/elixir/hamming/1/lib/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) do
    if length(strand1) !== length(strand2),
      do: {:error, "Lists must be the same length"},
      else: {:ok, count_differences(to_list(strand1), to_list(strand2))}
  end

  defp to_list(strand),
    do: String.graphemes(to_string(strand))

  defp count_differences(strand1, strand2)
       when is_list(strand1) and is_list(strand2),
       do:
         Enum.zip(strand1, strand2)
         |> Enum.reduce(_accumulator = 0, &count_differences/2)

  defp count_differences({strand1, strand2}, accumulator) do
    if strand1 === strand2,
      do: accumulator,
      else: accumulator + 1
  end
end

File: solutions/ocaml/hamming/1/hamming.ml
type nucleotide = A | C | G | T

let is_different nucleotide1 nucleotide2 =
  if nucleotide1 = nucleotide2
    then 0
    else 1

let rec compute_distance strand1 strand2 accumulated_distance =
  match strand1, strand2 with
  | [], [] ->
    Ok accumulated_distance
  | [], _ | _, [] ->
    Error "strands must be of equal length"
  | nucleotide1::tail1, nucleotide2::tail2 ->
      let distance_increment = is_different nucleotide1 nucleotide2 in
      let updated_distance = accumulated_distance + distance_increment in
      compute_distance tail1 tail2 updated_distance

let hamming_distance strand1 strand2 =
  compute_distance strand1 strand2 0